packages/jaeger-ui/src/model/trace-viewer.tsx

```diff
@@ -0,0 +1,40 @@
+// Copyright (c) 2017 Uber Technologies, Inc.
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+// http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+import { Span } from '../types/trace';
+
+type spansDict = { [index: string]: Span };
+
+// eslint-disable-next-line import/prefer-default-export
+export function getTraceName(spans: Span[]) {
+  const allTraceSpans: spansDict = spans.reduce((dict, span) => ({ ...dict, [span.spanID]: span }), {});
+  const rootSpan = spans
+    .filter(sp => {
+      if (!sp.references || !sp.references.length) {
+        return true;
+      }
+      const parentIDs = sp.references.filter(r => r.refType === 'CHILD_OF').map(r => r.spanID);
```

Comment:
What if you dropped filters for CHILD_OF? It's possible to have an async trace using FOLLOWS_FROM. I don't think the algorithm is actually sensitive to the type of reference. If anything,  I would filter on `r.spanContext.traceID` being the same as the current span, because a reference to an external trace is _definitely_ not a parent.

Reply:
Hm. Did I understand correctly, `span.references` **can't contain** refs to a children? I mean, if some span listed as a reference, it can be either a parent or a some missing ref, but not a child?
If so, I suggest to change this logic[ by this way](https://github.com/jaegertracing/jaeger-ui/pull/541/commits/9aa5b4bd17cd30b0ce655483aafc8e0c98e63038). There is no need to calculate a dict of all trace spans anymore, since `some()` can check traceID [by itself](https://github.com/jaegertracing/jaeger-ui/pull/541/commits/9aa5b4bd17cd30b0ce655483aafc8e0c98e63038#diff-2968d836188925a4a82346df3642203bR25). If there are no situations when `spans` (which is an argument to this function) contains more data then all `.references`, this approach looks more correct.

P. S. "I didn't find `spanContext` in a [types](https://github.com/jaegertracing/jaeger-ui/blob/master/packages/jaeger-ui/src/types/trace.tsx), think it should be `r.traceID` or `r.span.traceID` :)"

Reply:
Sorry, I was wrong.

There may be situations when there are references with same `traceID` **but** with incorrect (missing) `spanID`. It this case such span won't define as a root but it can be a root.

Updated a [logic](https://github.com/jaegertracing/jaeger-ui/pull/541/commits/8d3eb81a6d3b1d9f0dda577dde8d671326f821e5) and a [tests](https://github.com/jaegertracing/jaeger-ui/pull/541/commits/3b2cbae16293e78ef7ee8c04b3bfc80cc48e91ce). Now it should work correct.